crates/services/producer/src/block_producer.rs

```diff
@@ -219,9 +230,23 @@ where
                 .expect("It is impossible to overflow the current block height")
         });
 
-        let gas_price = gas_price
-            .or_else(|| self.gas_price_provider.gas_price(height.into()))
-            .ok_or(anyhow!("No gas price found for height {height:?}"))?;
+        let header = self.new_header(height, Tai64::now()).await?;
+
+        let consensus_params = self
+            .consensus_parameters_provider
+            .consensus_params_at_version(&header.consensus_parameters_version)?;
+
+        let gas_per_bytes = consensus_params.fee_params().gas_per_byte();
+        let max_gas_per_block = consensus_params.block_gas_limit();
+        let max_block_bytes = max_gas_per_block
+            .checked_div(gas_per_bytes)
+            .ok_or(anyhow!("Failed to calculated max block bytes"))?;
+
+        let gas_price = if let Some(inner) = gas_price {
+            inner
+        } else {
+            self.gas_price_provider.gas_price(max_block_bytes).await?
```

Comment:
It is better to move all calculations here=) Could you move the duplicated code into a small function that we could reuse in both places, please?=)

Plus, maybe for `dry_run`, we could just use the previous gas price without.